Update documentation regarding Kokkos-kernels data mapping (#886)
precice/precice.github.io: content/docs/configuration/basics/configuration-mapping.md

#### Execution backends

Since preCICE version 3, we integrated performance-portable mapping implementations into preCICE, allowing to compute mappings on different hardware backends.

> To use this feature, build preCICE from source with [Kokkos-Kernels or Ginkgo enabled](installation-source-dependencies.html). Both implementations rely additionally on Kokkos as a mandatory dependency. Overall, users target `CUDA`, `HIP`, `SYCL`, `OpenMP`, as backend. Note that preCICE inherits the backend from Kokkos at compile time. The `CPU` backend is the (default) regular backend in preCICE, which does not depend on Kokkos.

Reviewer: I don't get the sentence: 

> Overall, users target `CUDA`, `HIP`, `SYCL`, `OpenMP`, as backend

Could you reformulate?

> | Mapping tag             | Dependencies                                   | `cuda` | `hip` | `sycl` | `openmp` | `cpu` | MPI-parallelization |

Reviewer: ```suggestion
| Mapping tag             | Optional dependencies                                   | `cuda` | `hip` | `sycl` | `openmp` | `cpu` | MPI-parallelization |
```
??

Author: No. When configuring the depicted backend, the dependency is not optional?

Reviewer: True. If you manage to include Eigen in a reasonable way, then we don't need the optional here to distinguish between the two categories.

|-------------------------|------------------------------------------------|:------:|:-----:|:------:|:--------:|:-----:|---------------------|
| `rbf-pum-direct`        | —                                              | —      | —     | —      | —        | ✓     | distributed         |
| `rbf-pum-direct`        | Kokkos-Kernels and Kokkos (preCICE >= v3.4.1)  | ✓      | ✓     | ✓      | ✓        | —     | distributed         |
| `rbf-global-direct`     | Ginkgo and Kokkos (preCICE >= v3.2)            | ✓      | ✓     | —      | —        | —     | gather-scatter      |
| `rbf-global-direct`     | —                                              | —      | —     | —      | —        | ✓     | gather-scatter      |
| `rbf-global-iterative`  | Ginkgo and Kokkos (preCICE >= v3.2)            | ✓      | ✓     | ✓      | ✓        | —     | gather-scatter      |

> | `rbf-global-iterative`  | PETSc                                          | —      | —     | —      | —        | ✓     | distributed         |

Reviewer: Very briefly mention the mandatory Eigen? If someone looks at this table to find out what LinAlg backends we use?

To configure the executor, an additional subtag can be used in the mapping configuration:

```xml
<mapping:rbf-pumdirect direction="read" from="MyMesh2" to="MyMesh1" constraint="consistent">
 <basis-function:compact-polynomial-c6 support-radius="1.8"/>

>  <executor:cuda gpu-device-id="0"/>

Reviewer: Could it be better to use "auto" here? For users who copy from here.

Author: I would say no. ID zero should be available on all systems and using `auto` may not be appropriate for the global RBF system configuration.

</mapping:rbf-pum-direct>
```

> Mapping configurations, which follow a gather-scatter approach are always computes on a single MPI-rank, i.e., the mapping problem is gathered on the primary rank (potentially moved on the GPU), and then a solution is computed. Thus, mappings using a gather-scatter approach are not suitable for massively-parallel runs.

Reviewer: ```suggestion
Mapping configurations that follow a gather-scatter approach are always computed on a single MPI rank, i.e., the mapping problem is gathered on the primary rank (potentially moved to the GPU), and then a solution is computed. Thus, mappings using a gather-scatter approach are not suitable for massively-parallel runs.
```

> By contrast, distributed mappings solve their rank-local problem in parallel. In practice, this means that the configuration snippet above assigns multiple MPI ranks to a single GPU (with device ID 0). Since oversubscribing a device is typically undesired, use `gpu-device-id="auto"` to assign MPI ranks in a round-robin fashion to the available GPUs. Also note -- due to the different parallelization strategies -- `n-threads` configures the number of threads per executing rank, i.e., for a distributed parallelization, `n-threads=10` assigns 10 OpenMP threads to each MPI rank, whereas for a gather-scatter parallelization `n-threads=10` assigns 10 OpenMP threads to the primary rank only.

Reviewer: `n-threads` is not part of the example above. Could it be better to add another example?

Author: Added another example.

{% note %}
Note that as of preCICE v3.4 the `rbf-pum-direct` executor only supports `consistent` constraints. Supporting a `conservative` is [work in progress](https://github.com/precice/precice/pull/2536).
{% endnote %}

> More details on the feature can be found in [Schneider et al. 2023](https://doi.org/10.23967/c.coupled.2023.016).

Reviewer: We should probably also fine-tune this  sentence.

Author: The last sentence here or the `Note`?

Reviewer: The last sentence

Author: Would make sense to link to other documents, once they are available, I agree.
